bigtable/client/internal/row_reader.h

```diff
@@ -0,0 +1,158 @@
+  static int const NO_ROWS_LIMIT = 0;
+
+  RowReader(std::shared_ptr<DataClient> client, absl::string_view table_name,
+            RowSet row_set, int rows_limit, Filter filter,
+            std::unique_ptr<RPCRetryPolicy> retry_policy,
+            std::unique_ptr<RPCBackoffPolicy> backoff_policy);
+
+  using iterator = RowReaderIterator;
+
+  /**
+   * Input iterator over rows in the response.
+   *
+   * All iterators that do not point to end() are equivalent, that is,
+   * incrementing one iterator will increment all by advancing the
+   * stream.
+   */
+  iterator begin();
+
+  /// End iterator over the rows in the response.
+  iterator end();
+
+  /**
+   * Stop the read call and clean up the connection.
+   *
+   * Invalidates iterators.
+   */
+  void Cancel();
+
+ private:
+  /**
+   * Read and parse the next row in the response.
+   *
+   * Invalidates the previous row and feeds response chunks into the
+   * parser until another row is available.
+   *
+   * This call possibly blocks waiting for data.
+   */
+  void Advance();
+
+  /// Called by Advance(), does not handle retries.
+  grpc::Status AdvanceOrFail();
+
+  /**
+   * Move the index to the next chunk, reading data if needed.
+   *
+   * Returns false if no more chunks are available.
+   */
+  bool NextChunk();
+
+  /// Sends the ReadRows request to the stub.
+  void MakeRequest();
+
+  /// The input iterator returned by begin() and end()
+  class RowReaderIterator : public std::iterator<std::input_iterator_tag, Row> {
+   public:
+    RowReaderIterator(RowReader* owner, bool is_end)
+        : owner_(owner), is_end_(is_end) {}
+
+    RowReaderIterator& operator++();
+
+    Row const& operator*() { return owner_->row_.operator*(); }
+    Row const* operator->() { return owner_->row_.operator->(); }
+
+    bool operator==(RowReaderIterator const& that) const {
+      // All non-end iterators are equal.
+      return is_end_ == that.is_end_;
+    }
+
+    bool operator!=(RowReaderIterator const& that) const {
+      return !(*this == that);
+    }
+
+   private:
+    RowReader* const owner_;
+    bool is_end_;
+  };
+
+  std::shared_ptr<DataClient> client_;
+  absl::string_view table_name_;
+  RowSet row_set_;
+  int rows_limit_;
+  Filter filter_;
+  std::unique_ptr<RPCRetryPolicy> retry_policy_;
+  std::unique_ptr<RPCBackoffPolicy> backoff_policy_;
+
+  std::unique_ptr<grpc::ClientContext> context_;
+
+  std::unique_ptr<ReadRowsParser> parser_;
+  std::unique_ptr<
+      grpc::ClientReaderInterface<google::bigtable::v2::ReadRowsResponse>>
+      stream_;
+
+  /// The last received response, chunks are being parsed one by one from it.
+  google::bigtable::v2::ReadRowsResponse response_;
+  /// Number of chunks already parsed in response_.
+  int processed_chunks_;
+
+  /// Number of rows read so far, used to set row_limit in retries.
+  int rows_count_;
+
+  /// Holds the last read row, non-end() iterators all point to it.
```

Comment:
Alternatively, you can have the `end()` iterators have `owner_ == nullptr`, and move the Row to the actual iterators.  That would make them behave like other iterators do.

Reply:
done